Refactor package into one part dealing with LLVM and one part that builds a Vec on top of that

KristofferC opened the pull request:
This PR pretty much rewrites the package from scratch (with the exception of the indexing implemented by @tkf)  while keeping the API intact. The reason for this is that I felt that the code could gain a lot of clarity by clearly separating the parts that deal with LLVM/`llvmcall` and then build a SIMD `Vec` on top of that. The number of lines of code has also been reduced from ~1600 to ~1000 giving some support to this claim.

--------------------------------------

The code is structured as follows:

- `LLVM_intrinsics.jl` is pretty much a direct mapping of Julia Vectors (`NTuple{N, VecElement{T}}`) to the operators and intrinsics defined in https://llvm.org/docs/LangRef.html. It contains almost no higher level logic.
- `simdvec.jl` contains the `Vec` (wrapping the tuple of `VecElement`s) with definitions defined on it that maps to the intrinsics defined in `LLVM.jl`. In some cases, this is pretty automatic but in some cases requires some logic (like in the bitshifts partly to avoid undefined behavior or in the different conversions).
- `arrayops.jl` is the stuff that deals with Julia `Array` like `vload`, `vstore`, `vgather`.

-----------------------------------------

Things that have gotten added to the API:

- The `count_ones, count_zeros, leading_ones, leading_zeros, trailing_ones, trailing_zeros` family of functions.

- Type conversions and different types of reinterprets from scalar to vectors and back and between vectors of different size:

```jl
julia> v = Vec((Int32(2), Int32(4)))
<2 x Int32>[2, 4]

julia> reinterpret(Int64, v)
17179869186

julia> reinterpret(Vec{4, Int16}, v)
<4 x Int16>[2, 0, 4, 0]

julia> reinterpret(Vec{2, Int32}, 4)
<2 x Int32>[4, 0]

julia> convert(Vec{2, Float32}, v)
<2 x Float32>[2.0, 4.0]
```

- Uses the LLVM vector reduction intrinsics (https://llvm.org/docs/LangRef.html#experimental-vector-reduction-intrinsics) instead of a hand-rolled reducer. This requires LLVM 8 and thus Julia 1.4. It should be noted that these are marked as experimental.

-----------------------------------------

Things that has been removed from the API:

- Removed the `Val` arguments from many functions (`setindex`, `>>` etc). Julia's constant propagation + LLVM's optimization is enough for these not to be needed. Things are specialized on the constant just as well as if using `Val`.

- Removed the `Val{}` arguments and just use `Val()` consistently everywhere.

- Removed `exp10`. This used to just call `10^v` but the reason you would use `exp10` is that there is a more efficient implementation for it than the naive one. I feel that providing `exp10` gives a false impression that it provides a benefit over the naive version.

- Removed `all` on `Vec` of `Int`. There is no such correspondence to Julia numbers (`all` should operator on Bools).

-----------------------------------------

For the future, we should also think a bit how we could allow one to hook into the fast math flags defined in https://llvm.org/docs/LangRef.html#fast-math-flags. I guess we could try hook into the functionality provided by `@fastmath`.

I think a weak spot right now is all the different indexing. The combination of being able to use `Vec{N, T}`, `VecRange{N}` as the first argument, as well as the combination of alignments and non-temporal settings create a huge number of method combinations. In SIMD.jl many hundreds of lines are just defining similar methods with different orders of arguments and default values. Somehow the abstraction doesn't feel right here. I would want to at least make `VecRange{N}(i)` the default way to index because it feels unnecessary to have to pass the `T` in `Vec{N,T}` when it doesn't add any information.

Also, the `Aligned` flag sets the alignment to `N * sizeof(T)`. Can that really be right?

-----------------------------------------

Tagging some people that might be interested / can review. @tkf, @eschnett , @nlw0, @vchuravy, @chethega, @chriselrod 

Fixes https://github.com/eschnett/SIMD.jl/issues/65
Fixes https://github.com/eschnett/SIMD.jl/issues/54
Fixes https://github.com/eschnett/SIMD.jl/issues/51
Fixes https://github.com/eschnett/SIMD.jl/issues/20

vchuravy commented on src/LLVM_intrinsics.jl:
```diff
@@ -0,0 +1,530 @@
+for (fs, c) in zip([UNARY_INTRINSICS_FLOAT, UNARY_INTRINSICS_INT], [FloatingTypes, IntegerTypes])
+    for f in fs
+        @eval begin
+            @generated function $(f)(x::T) where T<:LT{<:$c}
```
Since you are already using `@eval` you can avoid the generated function here.
https://github.com/JuliaGPU/KernelAbstractions.jl/blob/50efa419a18947b9c277d235360be7e5919a8f6b/src/compiler/contract.jl#L14

eschnett replied:
The type `T` is not yet known in the `@eval`. To avoid the `@generated`, there needs to be an extra loop over all types `T`.

KristofferC replied:
You also need to loop over `N` though?

vchuravy replied:
Ah I see that particular bit of magic behind `LT` slipped past me. Sometimes I whish LLVM intrinsic would actually have multiple dispatch instead of just encoding it in a string. In theory it is supposed to be possible to have LLVM infer the type of the intrinsic from the arguments. but alas I haven't figure out how to make the work from the C interface.
https://llvm.org/doxygen/namespacellvm_1_1Intrinsic.html

vchuravy commented on src/LLVM_intrinsics.jl:
```diff
@@ -0,0 +1,530 @@
+################
+
+# These alignment numbers feels a bit dubious
+n_align(align, N, T) = align ? N * sizeof(T) : sizeof(T)
```
You probably want `max(align, Base.datatype_alignment(T)))`

chriselrod replied:
As is, `vmovntp` won't actually be generating streaming stores if `N` isn't a power of 2 (e.g., 56 byte alignment). That works.

eschnett replied:
Many CPU architectures are able to access memory more efficiently if the alignment is larger than what is required only for `T`. In the end, this is architecture-specific, but a good guess is that the alignment needs to be as large as the underlying vector type (not element type). Thus `N * sizeof(T)` is what I intended here. The unaligned case should probably be `Base.datatype_alignment(T)` instead of `sizeof(T)`:
```Julia
n_align(align, N, T) = align ? N * sizeof(T) : Base.datatype_alignment(T)
```

chriselrod replied:
I was confused here. Somehow I was thinking `Base.datatype_alignment(NTuple{7,Core.VecElement{Float64}})`.

eschnett replied:
That would be a good idea, but I don't think it works:
```Julia
julia> Base.datatype_alignment(NTuple{8, Core.VecElement{Float64}})
16
```
My CPU has AVX2 intrinsics, and I would have expected `64` as result. The reported alignment is for SSE2 intrinsics only.

chriselrod replied:
What about `sizeof`? Also, what Julia version are you on?

This computer is also AVX2 only:
```julia
julia> Base.datatype_alignment(NTuple{8, Core.VecElement{Float64}})
64

julia> versioninfo()
Julia Version 1.5.0-DEV.255
Commit dd426ba* (2020-02-09 18:28 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz
```

vchuravy replied:
> My CPU has AVX2 intrinsics, and I would have expected 64 as result. The reported alignment is for SSE2 intrinsics only.

Did you try on julia 1.4?  Before that 16 is the right answer... (heap alignment), on 1.4 I think we are finally supporting alignments bigger than the heap in more use-cases.

```
julia> Base.datatype_alignment(NTuple{8, Core.VecElement{Float64}})
64
```
this is on 1.5-dev.236

eschnett replied:
Yes, this is Julia 1.4:
```Julia
julia> Base.datatype_alignment(NTuple{8, Core.VecElement{Float64}})
16

julia> versioninfo()
Julia Version 1.4.0-rc1.0
Commit b0c33b0cf5 (2020-01-23 17:23 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin19.3.0)
  CPU: Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-8.0.1 (ORCJIT, skylake)
Environment:
  JULIA_MPI_PATH = /opt/local
  JULIA_MPI_LIBRARY_PATH = /opt/local/lib/openmpi-mp
```

Yes, an alignment of 2048 might cause problems, as it might waste a lot of memory. This should really be an alignment that leads to efficient memory access. Is there a way to query LLVM for the maximum vector size it will actually use for machine instructions? Or should we use the L1 cache line size to limit this value?

KristofferC replied:
https://github.com/JuliaLang/julia/pull/34517 is needed for that alignment to show 64.

KristofferC replied:
With `Base.datatype_alignment(T)` working like:

```
julia> Base.datatype_alignment(Vec{8, Float64})
64

julia> Base.datatype_alignment(Vec{16, Float64})
128

julia> Base.datatype_alignment(Vec{32, Float64})
256
```

in the upcoming 1.4, aren't we always in the `Aligned`-mode and the default should be changed from what is currently used in SIMD?

eschnett replied:
Vectorized code often has to interface with other, generic code that is not vectorized, by passing arrays forth and back. Arrays might also be multi-dimensional and will usually not be padded for a vectorized alignment. Thus the general case is that arrays are accessed in a non-aligned mode.

That is, these arrays are declared as arrays of scalars (e.g. `Matrix{Float64}`), and we're reading SIMD vectors from it. This is the signature for `vload` and `vstore`. If the array is of type `Matrix{Vec{8, Float64}}`, then it can make sense to assume it is aligned, depending on Julia's array allocator.

KristofferC replied:
Yeah, true.

vchuravy commented on src/arrayops.jl:
```diff
@@ -0,0 +1,284 @@
+using Base: Slice, ScalarIndex
+
+"""
+    ContiguousSubArray{T,N,P,I,L}
```
```suggestion
    ContiguousSubArray{T,N,P,I,L}

```

vchuravy commented on src/LLVM_intrinsics.jl:
```diff
@@ -0,0 +1,530 @@
+# LLVM operations and intrinsics
+module LLVM
```
I would prefer this being called `Intrinsics` or some such.

In the long run I would love some of the more complicated pieces of string manipulation to be converted to LLVM.jl
so the naming of this module would be conflicting, and for me it is most similar to `Core.Intrinsics`.
(There have been some discussions around removing the string based llvmcall from Base entirely, since it is not a stable API)

eschnett commented:
> Things that has been removed from the API:
> 
> Removed the Val arguments from many functions (setindex, >> etc). Julia's constant propagation + LLVM's optimization is enough for these not to be needed. Things are specialized on the constant just as well as if using Val.

I'm unsure about this, but I feel the burden would be on me to provide the test cases where `Val` provides a benefit.

> Removed the Val{} arguments and just use Val() consistently everywhere.

The `Val{}` were a design error and should now indeed be removed. I kept them for compatibility.

> Removed exp10. This used to just call 10^v but the reason you would use exp10 is that there is a more efficient implementation for it than the naive one. I feel that providing exp10 gives a false impression that it provides a benefit over the naive version.

`exp10` should be implemented in a way similar to `exp2` that is much more efficient than `10^`. There is a large performance difference. A simple implementation would be something like `exp10(x) = exp2(log2(10) * x)`. `exp2` is much faster than `pow` in libc.

eschnett commented:
Does this mean that Julia 1.4 will be required for these changes? I know many people who still use Julia 1.2, and in particular if there is also an API change, being backward compatible across multiple Julia versions would be a benefit.

eschnett commented:
I forgot the most important comment: Thanks for the work! I agree that this is the right direction.

eschnett commented on README.md:
```diff
@@ -63,7 +61,7 @@ ys1 = NTuple{4,Float32}(ys)
 
 # Update one element of a vector:
-ys = setindex(ys, 5, 3)   # cannot use ys[3] = 5
+ys = Base.setindex(ys, 5, 3)   # cannot use ys[3] = 5
```
There is no `setindex` in `Base`, there is only a `setindex!`.

chriselrod replied:
```julia
julia> Base.setindex
setindex (generic function with 2 methods)

julia> methods(Base.setindex)
# 2 methods for generic function "setindex":
[1] setindex(x::Tuple, v, i::Integer) in Base at tuple.jl:45
[2] setindex(nt::NamedTuple, v, idx::Symbol) in Base at namedtuple.jl:325

julia> versioninfo()
Julia Version 1.5.0-DEV.255
Commit dd426ba* (2020-02-09 18:28 UTC)
```

KristofferC replied:
I think there is one actually.

```
julia> Base.setindex
setindex (generic function with 2 methods)
```

eschnett replied:
Oops, I was confused.

KristofferC commented:
> Does this mean that Julia 1.4 will be required for these changes? I know many people who still use Julia 1.2, and in particular if there is also an API change, being backward compatible across multiple Julia versions would be a benefit.

The only thing that requires 1.4 is the new horizontal reduction intrinsics I think. So if we conditionally use the old hand-rolled one, it should be possible to support pre 1.4.

On the other hand, it isn't like the package was getting too much development (last functional change ~ a year ago) so those who use it in older Julia versions are probably just happy to keep using it as is and not to keen in upgrading packages at all.

chriselrod commented:
Regarding indexing API, what about `vload`/`vstore` methods that accept `AbstractArray{T,N}`s and `N`-length tuples of indices, using promotion to determine which method should be used?
If all indices are integers, it may be a scalar load.
If the `AbstractArray` is contiguous and column major, a `VecRange` as the first index, while all following indices are integers, would result in a llvm `load`/`store`.
The presence of `VecRange`s in any other position within the index-tuple, or the presence of `Vec{<:Any,<:Integer}` cause use of `gather`/`scatter`.

That's the API I settled on for LoopVectorization's code generation, since it means I can generate the same code largely independent of broader context (when it comes to loads/stores), and have multiple dispatch make it do the right thing.

Perhaps that is less desirable for code people are actively writing, because they will be more intentional about what they're doing; they're likely to know whether an index will be a scalar, `VecRange`, or a `Vec`.
But it may still help write more generic code in the future, where iterators may generate different sorts of indexes based on input arguments.

eschnett commented:
@KristofferC In principle yes, but given that Julia 1.4 isn't release yet, that would be a bit harsh. It would be nice (if it isn't too much effort) to give people a bit of time to upgrade (in the sense that Julia and SIMD can be upgraded separately).

eschnett commented:
@chriselrod I modelled the `vload` / `vstore` functions after OpenCL C. Since Julia has a much better type system than C, maybe there is room for improvement. Your suggestion sounds interesting.

KristofferC commented:
I need to fix the "codegen" for e.g. `Bool + Bool` because I haven't special cased that yet to do the `trunc` + `zext` dance yet. It is a bit unclear for me what the semantics for `Bool` should be. In Julia, `Bool`s immediately promote when used in arithmetic:

```
julia> true - true
0
```

and in some cases methods that are defined on `Int` are not defined on `Bool`:

```
julia> count_zeros(true)
ERROR: MethodError: no method matching count_ones(::Bool)
```

Should the just be considered an `i1` and then have all normal Int functions defined on them?
Or should we try to match Julia as much as possible. Any implicit promotion in SIMD.jl feels pretty pointless though.

eschnett commented:
In SIMD code, changing the vector width is often expensive. Some systems have therefore different bool sizes, such as `Bool32` and `Bool64`. They might internally represented very different from regular bools, i.e. they might even live in floating-point registers. I think SSE2 uses floating-point values with all-zero-bits and all-one-bits, PPC uses floating-point values that compare `>=0`, etc. Of course, more modern architectures (AVX512, CUDA) have special mask registers.

I don't know whether or how we can make LLVM generate efficient code here. I think the best would be to represent booleans as `i1`, and hope that LLVM knows about the efficient low-level SIMD representations.

I recommend against automatic conversion to `Int`, as `Int` would often have the wrong bit-width and thus lead to inefficient code. There is basically no automatic type conversion in SIMD.jl. I would also keep a distinction between `Bool` and `Int`, so even if the internal representation of `i1` and `i8` sounds similar, it doesn't have to be the same at a higher level. I don't see a point in defining `-(::Bool, ::Bool)` without automatic conversion and would omit it, similar for `count_zeros` etc.

We could define a type `Int1` (and I know cases where a vectorized `Int4` would be most convenient to have), but that's a different bag of chips.

KristofferC commented:
> I recommend against automatic conversion to Int, as Int would often have the wrong bit-width and thus lead to inefficient code. There is basically no automatic type conversion in SIMD.jl. I would also keep a distinction between Bool and Int, so even if the internal representation of i1 and i8 sounds similar, it doesn't have to be the same at a higher level. I don't see a point in defining -(::Bool, ::Bool) without automatic conversion and would omit it, similar for count_zeros etc.

Yeah, agreed. 

FWIW, I ran the tests on https://github.com/KristofferC/Tensors.jl (which uses SIMD.jl) and they passed. If anyone else is actually using SIMD.jl in their code, would be nice if you could run with this PR to see that it is non-breaking.

KristofferC commented:
- Implemented and added docs for the saturation intrinsics
- Implemented and added docs for the overflow intrinsics
- Extend Base.bitreverse (when it exists) since it got recently added to Julia
- Some other minor stuff

eschnett commented:
@KristofferC 

How do you want to continue with this PR? Is it ready to be merged?

Does the new version require Julia 1.5? If so, how do people usually continue to support older Julia versions? Would this be done on a branch? I'm not familiar with how to do that with the registration bot. Do you know?

KristofferC commented:
From my point of view, it is almost ready to be merged. The last feature I added runs into a bug on x86 (https://ci.appveyor.com/project/eschnett/simd-jl/builds/30989111/job/p0g8uhmkamy7ut8n#L56) which is similar to https://github.com/JuliaLang/julia/issues/29447. I think Julia needs to ship compiler-rt for it to work. I will  just disable that feature on X86 (overflow intrinsics)

This PR requires Julia 1.4 (not 1.5) but the overflow intrinsics are only available on 1.5.

Regarding wanting to support previous Julia versions, one often splits out a `release-2.0` branch from the commit before the Julia version requirement got updated and if bugfixes are needed for the previous Julia version those are made against that branch. Releases for 2.x can then be made from that branch at will.

KristofferC commented:
But before we tag, I want to go through the existing issues and see which ones are no longer relevant or see which ones can be easily fixed.

KristofferC commented:
From my point of view, this could be merged whenever. I still have a few tests to add (like making sure there are no spurious bounds checks inside `@inbounds`) but the PR is getting pretty big as it is.

chethega commented on src/LLVM_intrinsics.jl:
```diff
@@ -0,0 +1,647 @@
+
+###########
+# Bitcast #
+###########
```
We also need a trunc-and-bitcast intruction to represent vpmovmskb.

That would take `<n x i8>`, truncate it to `<n x i1>`, zero-pad it, and cast it to e.g. `i32` (or avx2). This can afaik not be reproduced with the other intrinsics here.

Likewise, we probably want an operation that takes `<n x i8>`, truncates to `<n x i1>` and `sext`s to `<n x i8>`;  possibly extending even more. This is in order to reduce mismatch between julia/llvm and x86 semantics (x86 tends to set all the bits, i.e. `sext`; julia has no representation of `<n x i1>`, and julia's Bool uses `zext` -style `<n x i8>`).

chethega commented on src/simdvec.jl:
```diff
@@ -0,0 +1,458 @@
+# See https://github.com/JuliaLang/julia/blob/7426625b5c07b0d93110293246089a259a0a677d/src/intrinsics.cpp#L1179-L1196
+# Shifting with a value larger than the number of bits in the type is undefined behavior
+# so set to zero in those cases.
+@inline function shl_int(x::Vec{N, T1}, y::Vec{N, T2}) where {N, T1<:IntegerTypes, T2<:IntegerTypes}
```
I think it would be much better to use llvm-semantics and document that they differ from julia semantics. People who use `SIMD.jl` care about speed, and can deal with the fact that shifts by more that width bits behave weirdly.

KristofferC replied:
Not "weirdly", it gives you a poison value which gives you undefined behavior in many cases. I don't think we should expose undefined behavior so easily from the `Vec` type. You can always just call `Intrinsics.shl`?

KristofferC replied:
We could also do the bitshifts modulo the size of the integer (which would match with e.g. rust). That would be cheaper since you can just `and` it.

eschnett replied:
Yes, this is a good idea. OpenCL does the same.

KristofferC replied:
What do we do about negative bitshifts?

eschnett replied:
Julia distinguishes between signed and unsigned bitshifts. For signed ones, it handles negative values as positive shifts in the opposite direction.

If we use the same function/operator name, we should use the same semantics. For full performance, people have to specify unsigned shift counts. If we think this is confusing, then we can reject signed shift counts.

KristofferC replied:
> If we use the same function/operator name, we should use the same semantics.

But in https://github.com/eschnett/SIMD.jl/pull/63#discussion_r385139094 we basically said that we would not keep the same semantics? In Julia, a shift larger than the bitsize does not wrap around but sets the value to zero which to me is a very different semantics.

eschnett replied:
Right. I thought Julia was undefined here, and we'd only be tightening things. But that's LLVM's semantics, not Julia's.

Should we keep Julia's semantics, and add new functions (`shl` / `shr`) that re-interpret as unsigned and wrap around?

In many cases, the shift count will be known at compile time, and in this case, things will be efficient anyway.

chethega replied:
So, I did some more reading on https://software.intel.com/sites/landingpage/IntrinsicsGuide/#techs=AVX,AVX2&cats=Shift

For hare-brained reason _mm256_slli_epi64 et al have julia semantics, while non-vectorized shifts ignore the upper bits. LLVM often fails to properly optimize code that enforces julia/AVX semantics; but if we emit x86_64 semantics, we have a mismatch with the AVX instructions we actually want. ARM does something else, because why not.

I see your point that the poison-value variant is slightly too poisonous for the default. So I guess we should document that poisonous `shl` is preferable to `<<` if known safe, and consider it our offering to ill-thought out llvm semantics.

eschnett replied:
I hereby volunteer you to write the doc string for these functions.

KristofferC commented:
I added `@fastmath` support in https://github.com/KristofferC/SIMD.jl/pull/1.

```jl
julia> f(a,b,c,d) = @fastmath a * b + c - d;

julia> v = Vec(1.0, 2.0, 3.0, 4.0);

julia> f(v, 1.0, v, 2.0)
<4 x Float64>[0.0, 2.0, 4.0, 6.0]

julia> @code_llvm debuginfo=:none f(v, 1.0, v, 2.0)
define void @julia_f_17933([1 x <4 x double>]* noalias nocapture sret, [1 x <4 x double>] 
...
  %5 = insertelement <4 x double> undef, double %2, i32 0
  %res.i = shufflevector <4 x double> %5, <4 x double> undef, <4 x i32> zeroinitializer
  %6 = getelementptr inbounds [1 x <4 x double>], [1 x <4 x double>] addrspace(11)* %1, i64 0, i64 0
  %7 = load <4 x double>, <4 x double> addrspace(11)* %6, align 16
  %8 = fmul fast <4 x double> %7, %res.i
  %9 = getelementptr inbounds [1 x <4 x double>], [1 x <4 x double>] addrspace(11)* %3, i64 0, i64 0
  %10 = load <4 x double>, <4 x double> addrspace(11)* %9, align 16
  %11 = insertelement <4 x double> undef, double %4, i32 0
  %12 = fsub fast <4 x double> <double -0.000000e+00, double undef, double undef, double undef>, %11
  %res.i1.neg = shufflevector <4 x double> %12, <4 x double> undef, <4 x i32> zeroinitializer
  %13 = fadd fast <4 x double> %8, %res.i1.neg
  %14 = fadd fast <4 x double> %13, %10
  %15 = getelementptr inbounds [1 x <4 x double>], [1 x <4 x double>]* %0, i64 0, i64 0
  store <4 x double> %14, <4 x double>* %15, align 32
  ret void
}
```

```jl
julia> @code_native debuginfo=:none f(v, 1.0, v, 2.0)
        .section        __TEXT,__text,regular,pure_instructions
        movq    %rdi, %rax
        vbroadcastsd    %xmm0, %ymm0
        vbroadcastsd    %xmm1, %ymm1
        vfmsub231pd     (%rsi), %ymm0, %ymm1 ## ymm1 = (ymm0 * mem) - ymm1
        vaddpd  (%rdx), %ymm1, %ymm0
        vmovapd %ymm0, (%rdi)
        vzeroupper
        retq
        nop

julia> g(a,b,c,d) =  a * b + c - d; # no @fastmath

julia> @code_native debuginfo=:none g(v, 1.0, v, 2.0)
        .section        __TEXT,__text,regular,pure_instructions
        movq    %rdi, %rax
        vbroadcastsd    %xmm0, %ymm0
        vmulpd  (%rsi), %ymm0, %ymm0
        vaddpd  (%rdx), %ymm0, %ymm0
        vbroadcastsd    %xmm1, %ymm1
        vsubpd  %ymm1, %ymm0, %ymm0
        vmovapd %ymm0, (%rdi)
        vzeroupper
        retq
        nopw    %cs:(%rax,%rax)
        nopl    (%rax,%rax)
```

KristofferC commented:
Added the fastmath commit to this PR.

tkf commented:
> Also, the `Aligned` flag sets the alignment to `N * sizeof(T)`. Can that really be right?

I just noticed that `Threads.Atomic` uses [`Base.gc_alignment`](https://github.com/JuliaLang/julia/blob/644753491de0d5d933c96e6f1f5876b3c7603cf4/base/reflection.jl#L358-L359). For example: https://github.com/JuliaLang/julia/blob/644753491de0d5d933c96e6f1f5876b3c7603cf4/base/atomics.jl#L347-L352

It calls `jl_gc_alignment` C API: https://github.com/JuliaLang/julia/blob/644753491de0d5d933c96e6f1f5876b3c7603cf4/src/julia_internal.h#L216-L236

Does it make sense to use `Base.gc_alignment(T)` when the alignment flag is `false`?

When it's `true`, it's basically SIMD.jl what it means to be "aligned" so I suppose it's OK? (Or was the question asking if it is "right" in terms of the performance?)

Alternatively, if `SIMD.Intrinsics` is meant to be a direct API to LLVM IR, maybe it can use `Val(integer)`? The default argument can still use `Base.gc_alignment(T)`.

tkf commented on src/arrayops.jl:
```diff
@@ -0,0 +1,285 @@
+
+@propagate_inbounds function vload(::Type{Vec{N, T}}, a::FastContiguousArray{T,1}, i::Integer, mask=nothing,
+                       ::Val{Aligned}=Val(false), ::Val{Nontemporal}=Val(false)) where {N, T, Aligned, Nontemporal}
+    @boundscheck checkbounds(a, i + N - 1)
```
```suggestion
    @boundscheck checkbounds(a, i:i + N - 1)
```

?

KristofferC replied:
Thanks, yeah

KristofferC commented:
Added some docs for the `@fastmath` integration.

KristofferC commented:
1.4 is now released. Any thoughts on how to progress here @eschnett

eschnett commented:
@KristofferC In general – once your pull request is ready, it should be applied. Are you referring to a particular question or suggestion (shifts, alignment, ...)?
